composer/2022_airflow_summit/data_analytics_process_expansion.py

```diff
@@ -0,0 +1,221 @@
+    # DF_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_NORMALIZED_TABLE_NAME}"
+    # PRCP_MEAN_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PRCP_MEAN_TABLE_NAME}"
+    # SNOW_MEAN_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_SNOW_MEAN_TABLE_NAME}"
+    # PHX_PRCP_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PHX_PRCP_TABLE_NAME}"
+    # PHX_SNOW_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PHX_SNOW_TABLE_NAME}"
+
+    # Create a SparkSession, viewable via the Spark UI
+    spark = SparkSession.builder.appName("data_processing").getOrCreate()
+    # Load data into dataframe if READ_TABLE exists
+    try:
+        df = spark.read.format("bigquery").load(READ_TABLE)
+    except Py4JJavaError as e:
+        raise Exception(f"Error reading {READ_TABLE}") from e
+
+    # filter out non-west states of the US
+    df = df.where((df.STATE == 'WA') | (df.STATE == 'OR') | (df.STATE == 'ID')
+                  | (df.STATE == 'MT') | (df.STATE == 'WY') | (df.STATE == 'CO')
+                  | (df.STATE == 'NM') | (df.STATE == 'AZ') | (df.STATE == 'UT')
+                  | (df.STATE == 'NV') | (df.STATE == 'CA'))
+    df.show(n=10)
+    print("After state filtering, # of rows remaining is:", df.count())
+
+    # filter out rows whose element value does not equal to PRCP or SNOW
+    df = df.where((df.ELEMENT == 'PRCP') | (df.ELEMENT == 'SNOW'))
+    df.show(n=10)
+    print("After elemet filtering, # of rows remaining is:", df.count())
+
+    # convert precipitation and snowfall from "tenths of a mm" to "mm"
+    @udf(returnType=FloatType())
+    def converter(val) -> float:
+        return val / 10
+    df = df.withColumn("VALUE", converter(df.VALUE))
+    df.show()
+
+    # return the year of each date
+    @udf(returnType=IntegerType())
+    def date_to_year(val) -> int:
+        return val.year
+
+    # extract the year of each date and rename it as YEAR
+    # This will allow us to merge the data based on the years they are created instead of date
+    df = df.withColumn("DATE", date_to_year(df.DATE)).withColumnRenamed('DATE', 'YEAR')
+
+    # each year's arithmetic mean of pricipation
+    prcp_mean_df = (
+        df.where(df.ELEMENT == 'PRCP')
+        .groupBy("YEAR")
+        .agg(avg("VALUE").alias("ANNUAL_PRCP_MEAN"))
+        .sort("YEAR")
+    )
+    print("prcp mean table")
+    prcp_mean_df.show(n=50)
+
+    # each year's arithmetic mean of snowfall
+    snow_mean_df = (
+        df.where(df.ELEMENT == 'SNOW')
+        .groupBy("YEAR")
+        .agg(avg("VALUE").alias("ANNUAL_SNOW_MEAN"))
+        .sort("YEAR")
+    )
+    print("snow mean table")
+    snow_mean_df.show(n=50)
+
+    # filter out the states to move on to the distance weighting algorithm (DWA)
+    annual_df = df.where(((df.STATE == 'CA') | (df.STATE == 'NV') | (df.STATE == 'UT')
+                          | (df.STATE == 'CO') | (df.STATE == 'AZ') | (df.STATE == 'NM')))
+
+    # latitude and longitude of Phoenix
```

Comment:
Why are you specifically including this?

Reply:
This information is needed to perform the distance weighting algorithm. I have also added a comment in the code